### PR TITLE
[InstCombine] Fold `(x == y) ? 0 : (x > y ? 1 : -1)` into `ucmp/scmp(x,y)`

### PR DESCRIPTION
**Pull request by Poseydon42**
Proofs: https://alive2.llvm.org/ce/z/Et9nrB

This also handles commuted cases of the same fold, with either the condition or the true/false values of the inner select being swapped.

### CODE REVIEW

**Comment by goldsteinn** [comment]
By inverse fold for you mean `x !- y ? x > y ? 1 : -1 : 0` or `x > y ? 1 : x == y ? 0 : -1`?

**Comment by nikic** [comment]
> By inverse fold for you mean `x !- y ? x > y ? 1 : -1 : 0` or `x > y ? 1 : x == y ? 0 : -1`?

It looks like this gets canonicalized: https://llvm.godbolt.org/z/7xTGM85G6 Probably good to add tests for that variant, but shouldn't need code changes.

**Comment by Poseydon42** [comment]
> By inverse fold for you mean `x !- y ? x > y ? 1 : -1 : 0` or `x > y ? 1 : x == y ? 0 : -1`?

The former; it gets canonicalized into `x == y ? 0 : (x > y ? 1 : -1)`, which is then handled by this patch.

**Comment by goldsteinn** [comment]
> > By inverse fold for you mean `x !- y ? x > y ? 1 : -1 : 0` or `x > y ? 1 : x == y ? 0 : -1`?
> 
> The former; it gets canonicalized into `x == y ? 0 : (x > y ? 1 : -1)`, which is then handled by this patch.

Any value in handling the latter?

**Comment by Poseydon42** [comment]
> > > By inverse fold for you mean `x !- y ? x > y ? 1 : -1 : 0` or `x > y ? 1 : x == y ? 0 : -1`?
> > 
> > 
> > The former; it gets canonicalized into `x == y ? 0 : (x > y ? 1 : -1)`, which is then handled by this patch.
> 
> Any value in handling the latter?

It's already being folded as well if I'm not mistaken.

File: llvm/test/Transforms/InstCombine/compare-3way.ll [hunk 1]
```diff
@@ -15,8 +15,7 @@ define void @test_low_sgt(i64 %a, i64 %b) {
 ; CHECK:       normal:
 ; CHECK-NEXT:    ret void
 ; CHECK:       unreached:
-; CHECK-NEXT:    [[EQ:%.*]] = icmp ne i64 [[A]], [[B]]
-; CHECK-NEXT:    [[RESULT:%.*]] = zext i1 [[EQ]] to i32
+; CHECK-NEXT:    [[RESULT:%.*]] = call i32 @llvm.scmp.i32.i64(i64 [[A]], i64 [[B]])
```
[the remark this answers is not in the document]
Comment (Poseydon42) [reply llvm/test/Transforms/InstCombine/compare-3way.ll:18]: I think for this particular case we can improve the way `ucmp`/`scmp` is handled in ConstraintElimination. The branch implies that within the `unreached` block `a >= b`, so we can replace the call with `icmp + zext/sext`. I don't think it's worth adding extra checks in InstCombine for this.

**Review by nikic — APPROVED** [review]
LGTM

@dtcxzyw Would be great if you could test this, in case it surfaces more optimization regressions...

**Comment by dtcxzyw** [comment]
> LGTM
> 
> @dtcxzyw Would be great if you could test this, in case it surfaces more optimization regressions...

In most cases, regressions occur when the result of `[u|s]cmp` is passes to a phi node, then the phi node is compared with a constant:
```
bb1:
%cmp = call i32 @llvm.ucmp(...)
br label %exit

exit:
%phi = i32 [%cmp, %bb1], [-1, %bb2]
%cmp2 = icmp slt i32 %phi, 0
```

Another missed optimization: if `ssat(X)` is only used to compare with zero, we can use `X` instead.
Alive2: https://alive2.llvm.org/ce/z/RZhWXF

**Comment by Poseydon42** [comment]
Rebased onto #107769. @dtcxzyw, can you please rerun the benchmark to see if the regression is fully gone. If that's the case I reckon we can go ahead and merge this and fix the other regression later as it seems to be less common.

**Comment by dtcxzyw** [comment]
See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1307/files#r1759649980. Can we remove the one-use check? Or just use `InVal->hasOneUser()` instead.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request merged**